pilot/pkg/model/sidecar.go

```diff
@@ -556,3 +515,42 @@ func (ilw *IstioEgressListenerWrapper) selectServices(services []*Service, confi
 	}
 	return filteredServices
 }
+
+func matchingServices(importedHosts []host.Name, service *Service, ilw *IstioEgressListenerWrapper) []*Service {
+	// If a listener is defined with port, we should match services with port.
+	// If an unix domain socket is given as a port, we should not match by port and include services based on hosts.
+	needsPortMatch := ilw.IstioListener != nil && ilw.IstioListener.Port != nil && ilw.IstioListener.Port.GetNumber() != 0
+	importedServices := make([]*Service, 0)
+
+	for _, importedHost := range importedHosts {
+		// Check if the hostnames match per usual hostname matching rules
+		if importedHost.Matches(service.Hostname) {
+			portMatched := false
+			if needsPortMatch {
+				for _, port := range service.Ports {
+					if port.Port == int(ilw.IstioListener.Port.GetNumber()) {
+						portMatched = true
+						break
+					}
+				}
+			} else {
+				importedServices = append(importedServices, service)
+				break
+			}
+			// If there is a port match, we should trim the service ports to the port specified by listener.
+			if portMatched {
+				for _, port := range service.Ports {
+					if port.Port == int(ilw.IstioListener.Port.GetNumber()) {
+						ports := []*Port{}
+						sc := service.DeepCopy()
+						ports = append(ports, port)
+						sc.Ports = ports
```

Comment:
This doesn't seem right? or maybe I just dont understand. It looks like `ports` is always equal to `port`. Won't we get the same svc duplicate multiple times with a single port now?

What I mean is a before we have `Service{ports: 1,2,3}` now we have `Service{ports: 1}, Service{Ports: 2}, Service{Ports: 3}`

Reply:
The partial services will be merged in to single service  here https://github.com/istio/istio/blob/master/pilot/pkg/model/sidecar.go#L225